File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    int[][] directions = new int[][] {{1,0}, {-1,0}, {0,1}, {0,-1}}; 
    public int numIslands(char[][] grid) {
        int result = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    result++;
                    dfs(grid, i, j);
                }
            }
        }
        return result;
    }

    public void dfs(char[][] grid, int r, int c) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length) return;
        if (grid[r][c] == '0') return;

        grid[r][c] = '0';
        for (int[] d: directions) dfs(grid, r + d[0], c + d[1]);
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-2.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        var time = 0
        val directions = listOf(Pair(0,1), Pair(0,-1), Pair(1,0), Pair(-1,0))
        val queue = ArrayDeque<Pair<Int,Int>>()
        val rows = grid.size
        val cols = grid[0].size
        var freshOrangesCount = 0
        for (i in 0..rows-1) {
            for (j in 0..cols-1) {
                if(grid[i][j] == 2) queue.add(Pair(i, j))
                if(grid[i][j] == 1) freshOrangesCount++
            }
        }

        while (queue.isNotEmpty()) {
            val level = queue.size
            var isRottenNextLevel = false
            repeat (level) {
                val (rr, rc) = queue.removeFirst()
                for ((r, c) in directions) {
                    val newRow = rr+r
                    val newCol = rc+c

                    if (newRow < 0 || newRow >= rows) continue
                    if (newCol < 0 || newCol >= cols) continue
                    if (grid[newRow][newCol] != 1) continue

                    grid[newRow][newCol] = 2
                    queue.addLast(Pair(newRow, newCol))
                    freshOrangesCount--
                    isRottenNextLevel = true
                }
            }

            if (isRottenNextLevel) time++
        }
        return if (freshOrangesCount > 0) -1 else time
        
    }
}
